# Самостоятельное задание 1 (+1 балл к оценке за курс).

In [ ]:
import pandas as pd

#### Описание данных

В папке Data находится информация о студентах. Всего 10 групп студентов. Файлы делятся на две категории:
    * Students_info_i - информация о студентах из группы i
    * Students_marks_i - оценки студентов из группы i за экзамены

### Одно из важных достоинств pandas $-$ это удобные методы реляционного взаимодействия с данными, аналогичные, например, возможностям SQL для слияния и конкатенации таблиц: merge, join, concat. Наличие готовых методов позволяет не реализовывать самостоятельно поэлементную обработку данных и оперировать сразу целыми таблицами данных.

Подробно об этих методах посмотрите тут: https://www.kaggle.com/residentmario/renaming-and-combining#Combining

#### Задание 1. Соберите всю информацию о студентах в одну таблицу df. В получившейся таблице должна быть информация и оценки всех студентов из всех групп. Напечатайте несколько строк таблицы для демонстрации результата.¶

In [ ]:
from glob import glob

In [ ]:
filenames_info = glob('/content/Students_info_*.csv')
filenames_marks = glob('/content/Students_marks_*.csv')
filenames_info.sort()
filenames_marks.sort()
df_info = pd.concat((pd.read_csv(file) for file in filenames_info), ignore_index = True)
df_marks = pd.concat((pd.read_csv(file) for file in filenames_marks), ignore_index = True)
df = pd.concat([df_info, df_marks], axis = 1)

In [ ]:
df.head()

index  gender race/ethnicity  ... math score reading score writing score
0      0  female        group B  ...         72            72            74
1      1  female        group C  ...         69            90            88
2      2  female        group B  ...         90            95            93
3      3    male        group A  ...         47            57            44
4      4    male        group C  ...         76            78            75

[5 rows x 11 columns]

#### Задание 2. Удалите столбец index у полученной таблицы. Напечатайте первые 10 строк таблицы.

In [ ]:
df = df.drop(columns = ['index'])

In [ ]:
df.head(10)

gender race/ethnicity  ... reading score writing score
0  female        group B  ...            72            74
1  female        group C  ...            90            88
2  female        group B  ...            95            93
3    male        group A  ...            57            44
4    male        group C  ...            78            75
5  female        group B  ...            83            78
6  female        group B  ...            95            92
7    male        group B  ...            43            39
8    male        group D  ...            64            67
9  female        group B  ...            60            50

[10 rows x 9 columns]

#### Задание 3. Выведите на экран размеры полученной таблицы

In [ ]:
df.shape

(1000, 9)

#### Задание 4. Выведите на экран статистические характеристики числовых столбцов таблицы (минимум, максимум, среднее значение, стандартное отклонение)

In [ ]:
import numpy as np
df.describe(include = [np.number])

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

#### Задание 5. Проверьте, есть ли в таблице пропущенные значения

In [ ]:
df.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
group                          0
math score                     0
reading score                  0
writing score                  0
dtype: int64

Ни в одном из столбцов нет NaN значений, а значит в таблице нет пропущенных значений.

#### Задание 6. Выведите на экран средние баллы студентов по каждому предмету (math, reading, writing)

In [ ]:
char = {'Score name': ['Math', 'Reading', 'Writing'], 
        'Mean': [df['math score'].mean(), df['reading score'].mean(), df['writing score'].mean()],}
df_char = pd.DataFrame(data = char) 
df_char

Score name    Mean
0       Math  66.089
1    Reading  69.169
2    Writing  68.054

**Задание 7. Как зависят оценки от того, проходил ли студент курс для подготовки к сдаче экзамена (test preparation course)? Выведите на экран для каждого предмета в отдельности средний балл студентов, проходивших курс для подготовки к экзамену и не проходивших курс.**

In [ ]:
df.groupby("test preparation course").mean()

math score  reading score  writing score
test preparation course                                          
completed                 69.695531      73.893855      74.418994
none                      64.077882      66.534268      64.504673

**Задание 8. Выведите на экран все различные значения из столбца lunch.**

In [ ]:
set(df["lunch"])

{'free/reduced', 'standard'}

**Задание 9. Переименуйте колонку "parental level of education" в 
"education", а "test preparation course" в "test preparation" с помощью метода pandas rename**
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html

In [ ]:
df = df.rename(columns={"parental level of education": "education", "test preparation course": "test preparation"})
df.head()

gender race/ethnicity  ... reading score writing score
0  female        group B  ...            72            74
1  female        group C  ...            90            88
2  female        group B  ...            95            93
3    male        group A  ...            57            44
4    male        group C  ...            78            75

[5 rows x 9 columns]

**Зафиксируем минимальный балл для сдачи экзамена**

In [ ]:
passmark = 50

**Задание 10. Ответьте на вопросы:**
    * Какая доля студентов сдала экзамен по математике (passmark > 50)?
    * Какая доля студентов, проходивших курс подготовки к экзамену, сдала экзамен по математике?
    * Какая доля женщин, не проходивших курс подготовки к экзамену, не сдала экзамен по математике? 

In [ ]:
(df["math score"] >= passmark).mean()

0.865

In [ ]:
(df[df['test preparation'] == 'completed']['math score'] >= passmark).mean()

0.9217877094972067

In [ ]:
(df[(df['test preparation'] == 'none') & (df['gender'] == 'female')]['math score'] < passmark).mean()


0.20958083832335328

**Задание 11. С помощью groupby выполните задания ниже. Также выведите время выполнения каждого из заданий.**
    * Для каждой этнической группы выведите средний балл за экзамен по чтению
    * Для каждого уровня образования выведите минимальный балл за экзамен по письму

In [ ]:
# загрузим библиотеку, чтобы посмотреть на различия во времени выполнения
import timeit

In [ ]:
start = timeit.default_timer()
print(df.groupby('race/ethnicity')['reading score'].mean())
stop = timeit.default_timer()
print('Time: ', stop - start)  

race/ethnicity
group A    64.674157
group B    67.352632
group C    69.103448
group D    70.030534
group E    73.028571
Name: reading score, dtype: float64
Time:  0.006767011000022194


In [ ]:
start = timeit.default_timer()
print(df.groupby('education')['writing score'].min())
stop = timeit.default_timer()
print('Time: ', stop - start) 

education
associate's degree    35
bachelor's degree     38
high school           15
master's degree       46
some college          19
some high school      10
Name: writing score, dtype: int64
Time:  0.006279146999986551


**Задание 12. Выполните задание 11 с помощью циклов. Сравните время выполнения.**

In [ ]:
start = timeit.default_timer()
a = set(df['race/ethnicity'])
for i in a:
  sum = 0
  count = 0
  for index, row in df.iterrows():
    if (row['race/ethnicity'] == i):
      sum += row['reading score']
      count += 1
  print(i, '\t', sum / count)
stop = timeit.default_timer()
print('Time: ', stop - start) 

group E 	 73.02857142857142
group D 	 70.03053435114504
group C 	 69.10344827586206
group A 	 64.67415730337079
group B 	 67.35263157894737
Time:  0.45376096700005064


In [ ]:
start = timeit.default_timer()
b = set(df['education'])
for i in b:
  cur_min = 0
  flag = 0
  for index, row in df.iterrows():
    if ((row['education'] == i) and ((flag == 0) or (cur_min > row['writing score']))):
      cur_min = row['writing score']
      flag = 1
  print(i, '\t', cur_min)
stop = timeit.default_timer()
print('Time: ', stop - start) 

bachelor's degree 	 38
some college 	 19
high school 	 15
some high school 	 10
master's degree 	 46
associate's degree 	 35
Time:  0.49216469899999993


Можно видеть, что первый и второй цикл выполняются значительно дольше, по сравнению с реализацией через функции библиотеки.

**Задание 13. Выведите на экран средние баллы студентов по каждому предмету в зависимости от пола и уровня образования. То есть должно получиться количество групп, равных 2 * (число уровней образования), и для каждой такой группы выыведите средний балл по каждому из предметов.**

Это можно сделать с помощью сводных таблиц (pivot_table):

https://www.kaggle.com/kamilpolak/tutorial-how-to-use-pivot-table-in-pandas

In [ ]:
grouped = pd.pivot_table(df, index = ['gender', 'education'])
grouped

math score  reading score  writing score
gender education                                                   
female associate's degree   65.250000      74.120690      74.000000
       bachelor's degree    68.349206      77.285714      78.380952
       high school          59.351064      68.202128      66.691489
       master's degree      66.500000      76.805556      77.638889
       some college         65.406780      73.550847      74.050847
       some high school     59.296703      69.109890      68.285714
male   associate's degree   70.764151      67.433962      65.405660
       bachelor's degree    70.581818      68.090909      67.654545
       high school          64.705882      61.480392      58.539216
       master's degree      74.826087      73.130435      72.608696
       some college         69.009259      64.990741      63.148148
       some high school     67.840909      64.693182      61.375000

#### Задание 14. Сколько студентов успешно сдали экзамен по математике?

Создайте новый столбец в таблице df под названием Math_PassStatus и запишите в него F, если студент не сдал экзамен по математике (балл за экзамен < passmark), и P иначе.

Посчитайте количество студентов, сдавших и не сдавших экзамен по математике.

Сделайте аналогичные шаги для экзаменов по чтению и письму.

In [ ]:
df['Math_PassStatus'] = df['math score'].apply(lambda x: 'F' if x < passmark else 'P')
df['Reading_PassStatus'] = df['reading score'].apply(lambda x: 'F' if x < passmark else 'P')
df['Writing_PassStatus'] = df['writing score'].apply(lambda x: 'F' if x < passmark else 'P')
df.head()

gender race/ethnicity  ... Reading_PassStatus Writing_PassStatus
0  female        group B  ...                  P                  P
1  female        group C  ...                  P                  P
2  female        group B  ...                  P                  P
3    male        group A  ...                  P                  F
4    male        group C  ...                  P                  P

[5 rows x 12 columns]

In [ ]:
print(df['Math_PassStatus'].value_counts())
print(df['Reading_PassStatus'].value_counts())
print(df['Writing_PassStatus'].value_counts())

P    865
F    135
Name: Math_PassStatus, dtype: int64
P    910
F     90
Name: Reading_PassStatus, dtype: int64
P    886
F    114
Name: Writing_PassStatus, dtype: int64


#### Задание 15. Сколько студентов успешно сдали все экзамены?

Создайте столбец OverAll_PassStatus и запишите в него для каждого студента 'F', если студент не сдал хотя бы один из трех экзаменов, а иначе 'P'.

Посчитайте количество студентов, которые сдали все экзамены.

In [ ]:
def f(row):
   if row['Math_PassStatus'] == 'F':
      return 'F'
   if row['Reading_PassStatus'] == 'F':
      return 'F'
   if row['Writing_PassStatus'] == 'F':
      return 'F'
   return 'P'

In [ ]:
df['OverAll_PassStatus'] = df.apply(lambda row: f(row), axis = 1)
df.head()

gender race/ethnicity  ... Writing_PassStatus OverAll_PassStatus
0  female        group B  ...                  P                  P
1  female        group C  ...                  P                  P
2  female        group B  ...                  P                  P
3    male        group A  ...                  F                  F
4    male        group C  ...                  P                  P

[5 rows x 13 columns]

In [ ]:
df['OverAll_PassStatus'].value_counts()

P    812
F    188
Name: OverAll_PassStatus, dtype: int64

То есть всего 812 студентов успешно сдали все экзамены.

#### Задание 16. Переведем баллы в оценки

### Система перевода баллов в оценки
####      90 и больше = A
####      80-89 = B
####      70-79 = C
####      60-69 = D
####      50-59 = E
####    меньше 50 = F (Fail)

Создайте вспомогательную функцию, которая будет по среднему баллу за три экзамена выставлять оценку студенту по данным выше критериям.

Создайте столбец Grade и запишите в него оценку каждого студента.

Выведите количество студентов, получивших каждую из оценок.

In [ ]:
def GetGrade(average_mark):
  if (average_mark >= 90):
    return 'A'
  if (average_mark >= 80):
    return 'B'
  if (average_mark >= 70):
    return 'C'
  if (average_mark >= 60):
    return 'D'
  if (average_mark >= 50):
    return 'E'
  return 'F'

Будем считать, что если студенты не справились с каким-то из экзаменов (passmark < 50), то независимо от среднего балла они получают в оценку F.

In [ ]:
df['Grade'] = df.apply(lambda row: 'F' if row['OverAll_PassStatus'] == 'F' else GetGrade((row['math score'] + row['reading score'] + row['writing score']) / 3), axis = 1)
df.head()

NameError: ignored

In [ ]:
df['Grade'].value_counts()